CLDSRV-402: optimize API

williamlardier opened the pull request:
For Artesca, more changes will be introduced with a bump of arsenal, and conflicts resolution (related to code not specific to this PR).

Federation & Integration (green) builds: https://github.com/scality/Integration/actions/runs/5309192284 & https://github.com/scality/Federation/actions/runs/5309145071

Context: The Veeam Ready tests for object with/without immutability are expecting us to perform a data deletion of 4 VM of 100GB each, splitted into 1MB files, in less than 4h. Today Artesca completes this in 10h, and RIng, 6hours.

The objective of this PR is to introduce optimizations for the MultiObjectDelete API:
* With a batching of the objects metadata reads, if the backends supports it (mongodb only for now).
* With an increased parallelism (that won't scale perfectly, but it helps dividing the latency for free)
* With a single call to the storage services, to avoid unecessary iowaits in the API.
* And finally, for Artesca only, supporting the direct deletion of the object metadata, if the associated bucket permits it (i.e. no notification/lifecycle configurations).

**Suggestion for the review**: by commit. Changes in the `lib/api/multiObjectDelete.js` file are not huge, but the diff might display new code, even if only indentation changed...

With these changes:
* Artesca completes the deletion in 39min instead of 10h.
* RING to be tested...

PR for 8.5: https://github.com/scality/cloudserver/pull/5205
8.6: https://github.com/scality/cloudserver/pull/5206
8.7: https://github.com/scality/cloudserver/pull/5207

Note 2: the integration changes are around all the breaking changes we have in this API across cloudserver versions, but the implementation is common between 7.x & 8.x.

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-402 contains:


* `8.7.24`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.22`

* `8.5.21`

* `8.6.10`

* `8.7.24`


Please check the `Fix Version/s` of CLDSRV-402, or the target
branch of this pull request.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.5
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

williamlardier commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.5
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

williamlardier commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.5
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

williamlardier commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.6
 $ git merge origin/w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.7
 $ git merge origin/w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

bert-e commented:
# History mismatch



Merge commit #788676cdc781079b42c1b2de09c4561f7502d083 on the integration branch
`w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api` is merging a branch which is neither the current
branch `improvement/CLDSRV-402-optimize-multiobjectdelete-api` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `improvement/CLDSRV-402-optimize-multiobjectdelete-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ghost commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.5
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

ghost commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.6
 $ git merge origin/w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

ghost commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.7
 $ git merge origin/w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

ghost commented:
/create_pull_requests

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-402-optimize-multiobjectdelete-api` into
`development/7.70`
* [w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api) will be merged into `development/8.5` (pull request #5205)
* [w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api) will be merged into `development/8.6` (pull request #5206)
* [w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api) will be merged into `development/8.7` (pull request #5207)



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_pull_requests**

jonathan-gramain commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+
+/**
+ * Initialization function for the MultiObjectDelete API that will, based on the
+ * current metadata backend,a ssess if metadata READ batching is supported. If
```
```suggestion
 * current metadata backend, assess if metadata READ batching is supported. If
```

jonathan-gramain commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+ * @param {object} log - logger object
+ * @return {undefined}
+ */
+function shouldBatchGetObjectsFromMetadata(bucketName, inPlay, log) {
```
It could be easier to read if the function returned a boolean, as one would expect from the function name, and just call the appropriate init function in the waterfall of `getObjectMetadataAndDelete`

williamlardier replied:
I would prefer keeping this function separately, that greatly ease the unit testing (and the base function is already quite complex :smile: ). Rather than just returning a flag, I will rename it `initializeMultiObjectDeleteWithBatchingSupport` instead, but I'm still open if you think we should change it anyway

williamlardier replied:
See https://github.com/scality/cloudserver/pull/5195/commits/9f7cc75b6259b61fcf0cbfa22d93fc544328f703

jonathan-gramain commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+    // transplants an additional piece of code that can greatly improve the API
+    // latency when the database supports batching.
+    if (backendSupportsBatching) {
+        const objectKeys = Object.keys(inPlay).map(entry => {
```
```suggestion
        const objectKeys = Object.values(inPlay).map(entry => {
```

And then use `entry` directly instead of `inPlay[entry]`

jonathan-gramain commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+    // in the API. This approach does not change the API architecture, but
+    // transplants an additional piece of code that can greatly improve the API
+    // latency when the database supports batching.
+    if (backendSupportsBatching) {
```
I would rather have this function just do `return backendSupportsBatching` (the function could then be called the same), and add the rest of the logic to populate the cache directly in the waterfall of `getObjMetadataAndDelete`, to avoid returning closures which can be confusing to read.

jonathan-gramain replied:
Sorry this kinda duplicates my earlier comment https://github.com/scality/cloudserver/pull/5195#discussion_r1231640057

jonathan-gramain commented on lib/metadata/metadataUtils.js:
```diff
@@ -84,6 +89,40 @@ function metadataGetObject(bucketName, objectKey, versionId, log, cb) {
+    // That is required with batching to properly map the objects
+    return metadata.getObjectsMD(bucketName, objects, log, (err, objMds) => {
+        if (err) {
+            log.debug('err getting object MD from metadata', { error: err });
```
```suggestion
            log.debug('error getting object MD from metadata', { error: err });
```

williamlardier replied:
Adding a 's' also to 'object' to differentiate with the other method.

jonathan-gramain commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                            // when versioning has been configured
+                            if (verCfg && !entry.versionId) {
+                                log.debug('trying to delete specific version ' +
+                                    ' that does not exist');
```
```suggestion
                                    'that does not exist');
```

jonathan-gramain commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                callback => callback(...processObjectVersion(entry, bucketName)),
+                // for obj deletes, no need to check acl's at object level
+                // (authority is at the bucket level for obj deletes)
+                (versionId, callback) => metadataGetObject(bucketName, entry.key,
```
I would expect this `metadataGetObject` call not to happen when the entry is already in cache (due to `metadataGetObjects` called earlier to populate the cache), but I don't see the mechanism to check this and skip `metadataGetObject`.

williamlardier replied:
I wonder if this is worth changing the API architecture for this? Originally this was something we discussed with @francoisferrand specifically not to change the API structure while benefiting from the previous cache. I am fine with both approaches, but the current way seem less risky...

We could just wrap this call and do something like:

```js
const handleObjectMetadata = (err, objMD, callback) => {
    // if general error from metadata return error
    if (err) {
        monitoring.promMetrics('DELETE', bucketName, err.code,
            'multiObjectDelete');
        return callback(err);
    }
    if (!objMD) {
        const verCfg = bucket.getVersioningConfiguration();
        // To adhere to AWS behavior, create a delete marker
        // if trying to delete an object that does not exist
        // when versioning has been configured
        if (verCfg && !entry.versionId) {
            log.debug('trying to delete specific version ' +
                ' that does not exist');
            return callback(null, objMD, versionId);
        }
        // otherwise if particular key does not exist, AWS
        // returns success for key so add to successfullyDeleted
        // list and move on
        successfullyDeleted.push({ entry });
        return callback(skipError);
    }
    if (versionId && objMD.location &&
        Array.isArray(objMD.location) && objMD.location[0]) {
        // we need this information for data deletes to AWS
        // eslint-disable-next-line no-param-reassign
        objMD.location[0].deleteVersion = true;
    }
    return callback(null, objMD, versionId);
};
...
(versionId, callback) => {
    if (cache && cache[`${entry.key}${versionId}`]) {
        return handleObjectMetadata(null, cache[`${entry.key}${versionId}`], callback);
    } else {
        return metadataGetObject(bucketName, entry.key, versionId, log, (err, objMD) => handleObjectMetadata(err, objMD, callback));
    }
...
```

jonathan-gramain replied:
Ah OK with the `cachedDocument` param, fine for me.

I don't know if it's the best approach but should work, personally I'd rather have abstracted the `metadataGetObjects` implementation itself into a single call to `getObjectsMD` or an async loop of `getObjectMD` calls (with concurrency limit), but if you want to minimize latency you may want to start processing results as soon as they arrive instead if the backend doesn't support `getObjectsMD`, however it's probably not a significant optimization in my sense.

francoisferrand replied:
c.f. my next comment: I guess this would be much clearer if the `callback` was kept as last param
Having the cache after the callback makes it very hard to read the call...

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.5`.

Please resolve the conflict on **the integration branch** (`w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ git pull  # or "git reset --hard origin/w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api"
 $ git merge origin/development/8.5
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```





*The following options are set:* **create_pull_requests**

ghost commented:
Note for reviewers: review commits will be integrated once I get more feedback on this PR. The implementation remains the same for integration branches.

jonathan-gramain commented on lib/services.js:
```diff
@@ -286,9 +286,12 @@ const services = {
      * @return {undefined}
      */
-    deleteObject(bucketName, objectMD, objectKey, options, log, cb) {
+    deleteObject(bucketName, objectMD, objectKey, options, log, cb, deferDeletion) {
```
Since `deferDeletion` refers to **data** locations deletion, it should I think be renamed for clarity, e.g. `deferLocationsDeletion`

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+ * @return {undefined}
+ */
+function initializeMultiObjectDeleteWithBatchingSupport(bucketName, inPlay, log) {
+    const backendSupportsBatching = constants.supportsBatchingMethods.includes(config.backends.metadata);
```
why do we need this `at all` ?

the support of batching is already handled in MetadataWrapper, which checks if `getObjects` is implemented : so here we should always initialize the cache by calling `MetadataWrapper.getObjectsMD`:
* If implemented, it will return the objects and fill the cache (or possibly return errors)
* If not implemented, it will simply create an empty cache
* Later on, the algorithm will check for object in the cache else call getObject, and we will be fine

williamlardier replied:
yes, I will remove it, at first the check in Arsenal was not here, now this check does not make sense anymore

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
     const skipError = new Error('skip');
     const objectLockedError = new Error('object locked');
+    let deleteFromStorage = [];
+    const initialStep = initializeMultiObjectDeleteWithBatchingSupport(bucketName, inPlay, log);
```
the callback initialized from the function makes the code harder to follow/read, and I don't see the benefit of the functional approach here : we may as well do what needs to be done in the waterfall, there be one "branch" anyway to handle both caching/no-caching...

williamlardier replied:
I reworked the function call to directly return the callback result: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR201 && https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR254

francoisferrand commented on lib/metadata/metadataUtils.js:
```diff
@@ -62,11 +62,16 @@ function getNullVersionFromMaster(bucketName, objectKey, log, cb) {
  * @return {undefined} - and call callback with err, bucket md and object md
  */
-function metadataGetObject(bucketName, objectKey, versionId, log, cb) {
+function metadataGetObject(bucketName, objectKey, versionId, log, cb, cachedDocument = null) {
```
shouldn't `cb` be the last param?

francoisferrand replied:
also IMO we should pass the whole cache to this function, and let it locate the object : if the caller already handles finding the object in the cache, then he may just as well skip the function altogether...

williamlardier replied:
> shouldn't cb be the last param?

We can make it a non optional parameter yes, and change all calls to this function...

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                // for obj deletes, no need to check acl's at object level
+                // (authority is at the bucket level for obj deletes)
+                (versionId, callback) => metadataGetObject(bucketName, entry.key,
+                    versionId, log, (err, objMD) => {
```
should just pass 'callback' and move this anonymous function as the next stage in the `async.waterfall`

williamlardier replied:
What is the rationale? That is not what was done before, and I do not want to change the structure...

williamlardier replied:
Done: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR260

francoisferrand replied:
rationale is really just that this is a waterfall call : whose call is to replace the "final" callback by the next stage in waterfall...

ie instead of:
```
metadataGetObject(bucketName, entry.key, versionId, log, (err, objMD) => {
    [...]
```
we should have:
```
metadataGetObject(bucketName, entry.key, versionId, log, callback),
(err, objMD) => {
    [...]
```

(unless there are other subtleties, there may be reasons why we cannot do this...)

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                    versionId, log, (err, objMD) => {
+                        // if general error from metadata return error
+                        if (err) {
+                            monitoring.promMetrics('DELETE', bucketName, err.code,
```
should the metric not be updated in the other error cases as well, e.g. in the final 'error' handling ?

williamlardier replied:
Added at the end of the function: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR424

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+ * @param {function} next - callback to call with error or decoded version
+ * @return {undefined}
+ **/
+function processObjectVersion(entry, bucketName) {
```
not clear what this does (and it certainly does not "process" a version): and especially when reading in the waterflow, extracting this function completely hides what is done...

maybe it should just be renamed to `decodeObjectVersion` ?

williamlardier replied:
Done: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR178

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -166,6 +169,60 @@ function _parseXml(xmlToParse, next) {
+            'null' : versionIdUtils.decode(entry.versionId);
+    }
+    if (decodedVersionId instanceof Error) {
+        monitoring.promMetrics('DELETE', bucketName, 404,
```
updating the `promMetrics` does not make sense here : this is not associated to the purpose of this function (decoding version id?), but to the overal process... and so should be kept somewhere in `getObjMetadataAndDelete` (eg keep level of abstraction)

williamlardier replied:
See above comment, it's now in the last callback, in case of error, I removed this one

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
-                            return callback(null, objMD, versionId);
+    return async.waterfall([
+        callback => initialStep((err, cache) => callback(err, cache)),
+        (cache, callback) => async.forEachLimit(inPlay, 50, (entry, moveOn) => {
```
the concurrency level should be a (global) constant at least, and/or maybe even a config parameter to give the possibly of (eventually) tweaking it...

francoisferrand replied:
> Shouldn't we do that in a separate PR for all the places we are relying on such constants

sure, if there are other such constant which should be cleaned'up, please create a `TechDebt` jira ticket

> I can do it for this one

yes please, this one should be fixed (the "constant" part at least) in this PR.
Fixing the rest of the world can wait, but at least we should improve the situation bit by bit when we find an issue and modify that code.

williamlardier replied:
I added support for that in the constants,; but also in the configuration file, so we can bind it with ZKOP at some point if needed
https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR255

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                        deleteInfo.deleted = true;
+                        if ((bucket.getLifecycleConfiguration && !bucket.getLifecycleConfiguration())
+                            && (bucket.getNotificationConfiguration && !bucket.getNotificationConfiguration())) {
+                            options.shouldOnlyDelete = true;
                         }
-                        if (error) {
-                            return callback(error);
+                        if (objMD.uploadId) {
```
this condition may change in the future : i find it should be isolated in a function so that we can easily evolve and maintain it (and possibly introduce more complex conditions)

williamlardier replied:
I created https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-b90cf333e787e31130d0434b2e48b75b69948c9e9c594c2d5f201058535b6ef4R1 that is also used when doing single deletes

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                    const deleteInfo = {};
+                    if (options && options.deleteData) {
+                        deleteInfo.deleted = true;
+                        if ((bucket.getLifecycleConfiguration && !bucket.getLifecycleConfiguration())
```
in case lifecycle, I think this would be needed only for _lifecycle transition to cold_ is used, and not even at the moment (we did not implement the delete event there event...) : not sure if we can/want to be more precise, or even just skip the test for now.

@nicolas2bert can you confirm if there are conditions where we need the full metadata in the oplog, for lifecycle?

francoisferrand replied:
in case lifecycle, I think this would be needed only for _lifecycle transition to cold_ is used, and not even at the moment (we did not implement the delete event there event...) : not sure if we can/want to be more precise, or even just skip the test for now.

@nicolas2bert can you confirm if there are conditions where we need the full metadata in the oplog, for lifecycle?

francoisferrand commented on lib/metadata/metadataUtils.js:
```diff
@@ -84,6 +89,40 @@ function metadataGetObject(bucketName, objectKey, versionId, log, cb) {
+ */
+function metadataGetObjects(bucketName, objectsKeys, log, cb) {
+    const options = { getDeleteMarker: true };
+    const objects = objectsKeys.map(objectKey => ({
```
there are already 2 maps in arsenal, and one more here : seems a lot, should maybe rework something to find a proper 'format' which suits all and avoids too many conversions...

francoisferrand commented on lib/services.js:
```diff
@@ -286,9 +286,12 @@ const services = {
      * @return {undefined}
      */
-    deleteObject(bucketName, objectMD, objectKey, options, log, cb) {
+    deleteObject(bucketName, objectMD, objectKey, options, log, cb, deferDeletion) {
         log.trace('deleting object from bucket');
```
`cb` should be the last param...

or actually, it seems to me we should not change the prototype of this function:
* `deleteObject(deferDeletion: true)` seems oxymoric, and not very readable
* it would be better instead of refactor with a function which "deletes the metadata and returns the list of locations" : which can be used to implement `deleteObject` as well as directly called by the multi delete process.
* maybe that intermediate function is not even needed, and we should simply call `metadata.deleteObjectMD` from the multi delete process.

williamlardier replied:
I ended up not changing the function at all, because the azure protection logic is related to the data services, and if we return all locations, we would end up deleting unnecessary things:
* it does not make sense to duplicate the `protectAzureBlocks` call
* yet if we call only a `deleteObjectMD` and `protectAzureBlocks` was supposed to return false, we change the behavior (even if this is harmless, this is not the objective of this PR)

I am sharing here the implementation I did before changing my mind:

```js
    /**
     * Deletes metadata of the object
     * @param {string} bucketName - bucket in which objectMD is stored
     * @param {string} objectKey - object key name
     * @param {object} options - other instructions, such as { versionId } to
     *                           delete a specific version of the object
     * @param {Log} log - logger instance
     * @param {function} cb - callback function
     * @return {undefined}
     */
    deleteMetadata(bucketName, objectKey, options, log, cb) {
        log.trace('Deleting object metadata and returning locations');
        assert.strictEqual(typeof bucketName, 'string');

        return metadata.deleteObjectMD(bucketName, objectKey, options, log,
            (err, res) => {
                if (err) {
                    return cb(err, res);
                }

                log.trace('deleteMetadata: metadata delete OK');
                return cb(err, res);
            });
    },

    /**
     * Deletes objects from a bucket
     * @param {string} bucketName - bucket in which objectMD is stored
     * @param {object} objectMD - object's metadata
     * @param {string} objectKey - object key name
     * @param {object} options - other instructions, such as { versionId } to
     *                           delete a specific version of the object
     * @param {Log} log - logger instance
     * @param {function} cb - callback from async.waterfall in objectGet
     * @return {undefined}
     */
    deleteObject(bucketName, objectMD, objectKey, options, log, cb) {
        log.trace('deleting object from bucket');
        assert.strictEqual(typeof bucketName, 'string');
        assert.strictEqual(typeof objectMD, 'object');
        const objGetInfo = objectMD.location;

        return services.deleteMetadata(bucketName, objectKey, options, log,
            (err, res) => {
                if (err) {
                    return cb(err, res);
                }
                log.trace('deleteObject: metadata delete OK');
                const deleteLog =
                    logger.newRequestLoggerFromSerializedUids(
                        log.getSerializedUids());
                if (objGetInfo === null) {
                    return cb(null, res);
                } else if (!Array.isArray(objGetInfo)) {
                    data.delete(objGetInfo, deleteLog);
                    return cb(null, res);
                }
                return data.batchDelete(objGetInfo, null, null,
                    deleteLog, err => {
                        if (err) {
                            return cb(err);
                        }
                        return cb(null, res);
                    });
            });
    },
```

francoisferrand commented on lib/services.js:
```diff
@@ -305,12 +308,16 @@ const services = {
+                        return cb(null, locations);
+                    }
+
+                    return data.batchDelete(locations, null, null, deleteLog, err => {
```
code used to call `data.delete` here instead of `batchDelete` : not sure we should change that... This is probably a minor optimisation, but still...

williamlardier replied:
Done https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-a76cecfcb0b408028fb3c7e02359a47730668d6424f8b02bb9522c162c21ce47R312

francoisferrand commented on tests/unit/api/multiObjectDelete.js:
```diff
@@ -185,3 +185,48 @@ describe('getObjMetadataAndDelete function for multiObjectDelete', () => {
             });
     });
```
missing tests here, esp. to verify the behaviours which you changed

* Is the cache populated when supported (e.g. call to `getObjectsMD`)
* Is the cache used (e.g. `getObjectMD` not called when object is already in cache)
* Is it working in both "with cache" and "without cache" modes
* Proper "batch" error handling : e.g. verify that we (batch) delete the `data` for each `metadata` we deleted even if there are errors on "other" objects

williamlardier replied:
I updated some unit tests: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-c2cf43a7212798ef328432b312d2ef420db8c78323dad28d5ea930cdb251b519R227
and added a new test for the last point you mentioned: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-c2cf43a7212798ef328432b312d2ef420db8c78323dad28d5ea930cdb251b519R195

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-402 contains:


* `7.70.22`

* `8.5.21`

* `8.6.10`

* `8.7.25`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.22`

* `8.6.10`

* `8.7.25`

* `8.8.0`


Please check the `Fix Version/s` of CLDSRV-402, or the target
branch of this pull request.




*The following options are set:* **create_pull_requests**

francoisferrand commented:
I wonder if we can/should introduce some flag to "disable" the optimization by default, and allow us more control when we enable it ; esp. since we are introducing this in older waterfall branch and may need to have it in patch versions of product...

in theory it should be added, but on the other hand I am not sure it is "easy" to disable (i.e. the cost and risk of allowing deactivation may not be worth the benefit) or if this is a critical/blocking fix for product (which we need to enable ASAP everywhere, and which should not ever be disabled).

ghost commented:
Rebased following CI fixes merged in another branch.

francoisferrand commented on lib/Config.js:
```diff
@@ -1320,6 +1321,18 @@ class Config extends EventEmitter {
+        this.multiObjectDeleteConcurrency = constants.multiObjectDeleteConcurrency;
+        if (!isNaN(config.multiObjectDeleteConcurrency)) {
+            const extractedNumber = Number.parseInt(config.multiObjectDeleteConcurrency, 10);
+            if (extractedNumber > 0 && extractedNumber < 1000 && extractedNumber % 1 === 0) {
```
> `extractedNumber % 1 === 0`

i don't see the point of this check, `parseInt` returns an integer anyway

francoisferrand commented on lib/Config.js:
```diff
@@ -1320,6 +1321,18 @@ class Config extends EventEmitter {
 
         this.lifecycleRoleName = config.lifecycleRoleName || null;
+        this.multiObjectDeleteConcurrency = constants.multiObjectDeleteConcurrency;
+        if (!isNaN(config.multiObjectDeleteConcurrency)) {
```
what is the point of checking for `NaN` here: `Number.parseInt` will handle it and return an error/NaN when needed anyway?

francoisferrand commented on lib/Config.js:
```diff
@@ -1320,6 +1321,18 @@ class Config extends EventEmitter {
+        }
+
+        this.multiObjectDeleteEnableOptimizations = true;
+        if (config.multiObjectDeleteEnableOptimizations === false) {
```
```suggestion
        if (!config.multiObjectDeleteEnableOptimizations) {
```

williamlardier replied:
This was kept unchanged, because we want to enable the optimization by default, and this change would disable it, as there is no such configuration by default.

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -242,114 +286,152 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                const deleteLog = logger.newRequestLoggerFromSerializedUids(log.getSerializedUids());
+                return data.batchDelete(deleteFromStorage, null, null, deleteLog, err => {
+                    if (err) {
+                        log.error('error deleting objects from data backend', { error: err });
```
when this happens, all the "metrics" computed above may be wrong : numOfObjectsRemoved, successfullyDeleted, totalContentLengthDeleted...

* if `batchDelete` works somewhat "atomically" (e.g. it succeeds to delete all OR fails and delete nothing), then the logic can be tweak to count properly;
* otherwise, we may need that function to return the "list" of objects succesfully deleted (and their size?), so we can do the appropriate count

williamlardier replied:
The current behavior is that, in case of any error, the function returns, and the metrics are unchanged, as they are only computed based on (successful) operations on the metadata, not the storage. A failure in deleting the actual data would lead to the whole API to fail. The only difference I see here is the number or orphans we would get, more than an issue with the metrics. When the API fails, we do not push any metric... So to me, this is not something to handle, as we push no metrics anyway.

The main question is more about handling or orphans: this can be mitigated by pre-splitting the array from `1000` to `1000/${multiObjectDeleteConcurrency}` as this will be aligned with the current parallelism, without the drawback of waiting and doing the calls one by one. I will make the associated changes.

KazToozs reviewed (approved):
lgtm

Seragonia committed:
CLDSRV-402: optimize multideleteobject API

- Parallelism is increased to reduce the latency
- If the backend supports it, batching is used
- Batch the deletion of objects from storage
- Flag to disable or enable the optimization, as well as
  a way to tune it.

Seragonia committed:
CLDSRV-402: test multi object delete optimization

Seragonia committed:
CLDSRV-402: bump arsenal dependency

Seragonia committed:
CLDSRV-402: bump project version

williamlardier commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_pull_requests**

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-402 contains:


* `7.70.22`

* `8.5.21`

* `8.6.10`

* `8.7.25`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.22`

* `8.6.10`

* `8.7.25`

* `8.8.0`


Please check the `Fix Version/s` of CLDSRV-402, or the target
branch of this pull request.




*The following options are set:* **create_pull_requests**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.6
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```





*The following options are set:* **create_pull_requests**

williamlardier commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.7
 $ git merge origin/w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```





*The following options are set:* **create_pull_requests**

bert-e commented:
# Unknown command



I didn't understand this comment by @williamlardier:

> /force_delete

I don't know what `force_delete` means.

Please **edit** or **delete** the corresponding comment so I can move on.




*The following options are set:* **create_pull_requests**

ghost commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_pull_requests**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.6
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```





*The following options are set:* **create_pull_requests**

ghost commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api origin/development/8.7
 $ git merge origin/w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```





*The following options are set:* **create_pull_requests**

ghost commented:
ping

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-402-optimize-multiobjectdelete-api` into
`development/7.70`
* [w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api) will be merged into `development/8.6` (pull request #5232)
* [w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/improvement/CLDSRV-402-optimize-multiobjectdelete-api) will be merged into `development/8.7` (pull request #5233)
* [w/8.8/improvement/CLDSRV-402-optimize-multiobjectdelete-api](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-402-optimize-multiobjectdelete-api) will be merged into `development/8.8` (pull request #5234)



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_pull_requests**

Seragonia committed:
CLDSRV-402: bump project version

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api` with contents from `improvement/CLDSRV-402-optimize-multiobjectdelete-api`
and `development/8.6`.

Please resolve the conflict on **the integration branch** (`w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ git pull  # or "git reset --hard origin/w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api"
 $ git merge origin/development/8.6
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/CLDSRV-402-optimize-multiobjectdelete-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.6/improvement/CLDSRV-402-optimize-multiobjectdelete-api
```





*The following options are set:* **create_pull_requests**

ghost commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`

* :heavy_check_mark: `development/8.8`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, create_pull_requests**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`

* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`



Please check the status of the associated issue CLDSRV-402.

Goodbye williamlardier.
